### PR TITLE
`azurerm_application_insights` - Disable Rule and Action Group that are automatically created when creating this resource

### LINKED ISSUE #10563: r/application_insights: Terraform should remove the automatically created nested items post-creation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

This issue is a direct followup to https://github.com/terraform-providers/terraform-provider-azurerm/issues/8812 as I think you completly missed the point of that issue.

Whenever a new insights resource is created via `azurerm_application_insights` azure automatically creates an alert rule called "Failure Anomalies" with the condition "Whenever the Failure Anomalies criteria is met" and an action group "Application Insights Smart Detection" which is also automatically created. (NB: It takes several minutes for this rule to show up in the portal after insights is created)

As Roshan B pointed out this creates a high volume of undesired alert messages sent to the wrong people.

What Roshan B wanted in his issue was the ability to turn off the creation of that alert rule

> So it would be great if we could get a way to disable these alerts in azurerm_application_insights resource.

What you instead delivered in your fix was the option to create more of them. 😮 

To make things worse, that "Failure Anomalies" rule and "Application Insights Smart Detection" alert group are completly outside of terraforms control. They are not accessible as a resource and thus cannot easily be altered (like changing the action group). You have to fiddle with az CLI call and/or ARM templates to get a grip on that alert rule. This is far from the stable automation process I'd like to achieve with terraform.

### New or Affected Resource(s)

* azurerm_application_insights
* azurerm_application_insights_smart_detection_rule

### Potential Terraform Configuration

```hcl
resource "azurerm_application_insights" "shared" {
  name                = "randd-fakeapp-ai"
  location            = var.location
  resource_group_name = var.resource_group_name
  application_type    = "other" # 'other' means that a general instance is to be created
  tags                = var.tags

  auto_create_failure_anomalies_rule = false   <<<< some option like this would be nice
}
```

### References

https://docs.microsoft.com/en-us/azure/azure-monitor/app/proactive-failure-diagnostics#how-it-works

### PR DESCRIPTION
**Pull request by mbfrahry**
Fixes #10563

### CODE REVIEW

**Review by katbyte — APPROVED** [review]
LGTM 🦀

File: internal/services/applicationinsights/application_insights_resource.go [hunk 1]
```diff
@@ -274,7 +277,67 @@ func resourceApplicationInsightsCreateUpdate(d *pluginsdk.ResourceData, meta int
...
 	}
 
-	d.SetId(resourceId)
+	// https://github.com/hashicorp/terraform-provider-azurerm/issues/10563
+	// Azure creates a rule and action group when creating this resource that are very noisy
+	// We would like to delete them but deleting them just causes them to be recreated after a few minutes.
+	// Instead, we'll opt to disable them here
+	if d.IsNewResource() && meta.(*clients.Client).Features.ApplicationInsights.DisableGeneratedRule {
+		err := pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
+			time.Sleep(30 * time.Second)
```
Comment (tombuildsstuff) [inline internal/services/applicationinsights/application_insights_resource.go:286]:
```suggestion
		// TODO: replace this with a StateWaitFunc
		err := pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
			time.Sleep(30 * time.Second)
```

File: internal/services/applicationinsights/application_insights_resource.go [hunk 2]
```diff
@@ -274,7 +277,67 @@ func resourceApplicationInsightsCreateUpdate(d *pluginsdk.ResourceData, meta int
...
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
+			time.Sleep(30 * time.Second)
```
Comment (tombuildsstuff) [inline internal/services/applicationinsights/application_insights_resource.go:313]: ditto here:

```suggestion
		// TODO: replace this with a StateWait func
		err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
			time.Sleep(30 * time.Second)
```

File: internal/services/applicationinsights/application_insights_resource.go (hunk 1 above)
Comment (tombuildsstuff) [inline internal/services/applicationinsights/application_insights_resource.go:286]: we shouldn't include Sleeps here, adding a TODO to clean this up post-3.0:

```suggestion
		// TODO: replace this with a StateWait func
		err = pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), func() *pluginsdk.RetryError {
			time.Sleep(30 * time.Second)
```

File: internal/provider/features.go [hunk 3]
```diff
@@ -25,6 +25,20 @@ func schemaFeatures(supportLegacyTestSuite bool) *pluginsdk.Schema {
...
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"disable_generated_rule": {
+						Type:     pluginsdk.TypeBool,
+						Optional: true,
+					},
+				},
+			},
+		},
```
Comment (tombuildsstuff) [inline internal/provider/features.go:40]: this needs to be added to the Features page, and also to the complete Example usage there (highlighting the default value of `false`): https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/features-block

**Review by tombuildsstuff — APPROVED** [review]
Three comments but this otherwise LGTM 👍

File: website/docs/guides/features-block.html.markdown [hunk 4]
```diff
@@ -31,6 +31,10 @@ provider "azurerm" {
       purge_soft_delete_on_destroy = true
     }
 
+    application_insights {
+      disable_generated_rule = true
```
Comment (tombuildsstuff) [inline website/docs/guides/features-block.html.markdown:35]: can we update this to the default value:

```suggestion
      disable_generated_rule = false
```

File: internal/provider/features.go (hunk 3 above)
Comment (mbfrahry) [reply internal/provider/features.go:40]: Fixed!

**Review by tombuildsstuff — APPROVED** [review]
LGTM 👍

**Comment by tombuildsstuff** [comment]
Build failure fixes in #15952

**Comment by michelefa1988** [comment]
@tombuildsstuff would you know when this version will be available?

**Comment by sebader** [comment]
It would be super helpful if this toggle was mentioned on the doc for the Application Insights resource. I just found this by chance here, looking at the features docs page is not the most natural place for people to check.

https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/application_insights

**Pull request merged**